spec.md

```diff
@@ -583,6 +583,20 @@ message VolumeInfo {
 }
 ```
 
+##### CreateVolume Errors
+
+If the plugin is unable to complete the CreateVolume call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code.
+The CO MUST implement the specified error recovery behavior when it encounters the gRPC error code.
+
+Condition | gRPC Code | Description | Recovery Behavior
+| --- | --- | --- | --- |
+| Volume already exists | 6 ALREADY_EXISTS | Indicates that a volume corresponding to the specified volume `name` already exists. Plugin MUST also return a valid `CreateVolumeResponse`. | Caller MUST assume the `CreateVolume` call succeeded. |
```

Comment:
I don't think this is possible, at least for golang-based servers. Has anyone tested this approach? I'm looking at the golang grpc code and it seems like if an RPC generates an error, then the xxxResponse is dropped and just "status" is returned to the client: https://github.com/grpc/grpc-go/blob/8ff8683602df4fb6a85c22f2da944bfc8fc73c50/server.go#L877

Reply:
clarification: xxxResponse being the non-status payload

Reply:
I believe you do this with `status.New(code, msg).WithDetails(proto)`... and then `s.Err()` to get the underlying `error` implementation.

Reply:
It seems a response with a a non-OK status is not recommended from the discussion of the open issue here: https://github.com/grpc/grpc/issues/12824#issuecomment-335624435

Reply:
In the example above the detailed error is encoded with the status message and can be decoded on the other side (or at least that's how I understand the API, I haven't tried to use it yet).

Reply:
See the GoCSI PR I linked. I’m already using the error message there.

Reply:
That was my understanding also, although in my case it was based on hearsay.  Since we claim that these calls are idempotent, I think it would be OK if we just returned no error for the case where the input options are all the same and the volume is already created.  That would allow us to maintain idempotency.

We could then treat ALREADY_EXISTS as an error case, for the instance when someone tries to create the same volume with different inputs.

Reply:
Although the current gRPC implementation allows us to return a response with a non-OK status, it doesn't seem a reliable feature to me, as the convention described in the comment I referenced is against that. Serializing `VolumeInfo` in the error details sounds more reliable, but the CO would have to deserialize it manually, and this should be explained in the spec.

Reply:
Hi @chhsia0,

It actually works fine as is. Please see GoCSI's [idempotent create tests](https://github.com/akutz/gocsi/blob/b356efea1c3729a5fd50455cb592e12314353083/controller_test.go#L147-L256) where I scale up to 100,000 concurrent `CreateVolume` RPCs, and they all return the `VolumeInfo` and gRPC error successfully.

Reply:
I've read your PR and I'm sure your tests work just fine. I raise this question because it seems to me not a good idea to rely on a gRPC implementation-dependent feature that is not ensured by gRPC's semantics, and I was wondering if this should be a concern, or someone can tell me that I'm totally wrong and the support of response with a non-OK status is guaranteed by gRPC's community.

Reply:
Hi @chhsia0,

> to rely on a gRPC implementation-dependent feature that is not ensured by gRPC's formal semantics

I'm not sure what you mean? Perhaps I'm missing something, but how is returning a tuple with both fields populated any less formal than the gRPC error codes? Or is the ability to return both a non-nil response **and** error not actually a part of gRPC? Is it purely a side-effect of Go's implementation?

Reply:
According Dr. Kumar's (I think he's the engineering lead of gRPC) comment I cross-referenced above, the convention of gRPC's protobuf-based type-safe API layer for unary response is that "Client side SHOULD drop any response messages received with a non-OK response."

However, he also mentioned that gRPC's wire protocol does allow any combination. My interpretation is that the wire protocol does accept a non-OK status with a response, but proto-based API designs such as CSI should not rely on this. @saad-ali can you clarify on this? Thanks.

Reply:
Hi @chhsia0,

>  "Client side SHOULD drop any response messages received with a non-OK response."

If the above is in fact true then I agree that the error details is the correct location for any idempotent data.

Reply:
Not sure if this has been discussed, but another possible option (if we want to avoid using the error details for any reason) is to return an OK and introduce a boolean field in the response to indicate if a volume of the same capabilities/parameters has been created before, and use `ALREADY_EXISTS` as an error suggested by @julian-hj.

Reply:
Hi @chhsia0,

I don't like that approach because it subverts (or compliments in a way that replaces) the existing error model that is designed to handle the case of something that already exists. I recommend using the gRPC error details as you've already suggested.

Reply:
As far as I understand it, if a function is idempotent, then calling it twice with the same inputs should not generate an error code.  So if "ALREADY_EXISTS" is an error, we should not use it for the case where the client just got impatient and repeated an RPC call, or repeated a call after a dropped connection.

What I am not clear on is whether "ALREADY_EXISTS" is deemed as an error condition or not?

Reply:
I'm fine either way.
For me idempotent means you can run the same thing multiple times without side-effects. The action can be idempotent (nothing changed) while still providing more information to the caller (this thing already exists) instead of "yep we did that".

However there is a case where a volume could exist but the state of it conflicts with the passed in options. In which case ALREADY_EXISTS could work... but also maybe a different error code applies to this.